src/core/demangle_cxx.d

```diff
@@ -0,0 +1,152 @@
+module core.demangle_cxx;
+
+version (OSX)
+    enum libCXXABI = "libc++abi.dylib";
+else version (linux)
+    enum libCXXABI = "libstdc++.so";
+else
+    enum libCXXABI = null;
+
+// from https://gcc.gnu.org/onlinedocs/libstdc++/libstdc++-html-USERS-4.3/a01696.html#76957f5810098d2ffb5c62f43eda1c6d :
+/+
+0: The demangling operation succeeded.
+-1: A memory allocation failiure occurred.
+-2: mangled_name is not a valid name under the C++ ABI mangling rules.
+-3: One of the arguments is invalid.
++/
+extern (C) pure nothrow char* __cxa_demangle(const char* mangled_name,
+        char* output_buffer, size_t* length, int* status);
+
+struct DemangleCXX
+{
+    typeof(__cxa_demangle)* fun_cxa_demangle;
+    char[] buffer;
+    // allocated via malloc (see __cxa_demangle)
+    char[] buffer2;
+
+    this(int unused) nothrow
+    {
+        fun_cxa_demangle = get_cxa_demangle;
+    }
+
+    ~this()
+    {
+        // CHECKME: is that safe to do? (cf caveats in https://github.com/dlang/druntime/pull/1836)
+        import core.memory : pureFree;
+
+        pureFree(buffer2.ptr);
+    }
+
+    void reserve(size_t length) pure nothrow
+    {
+        // NOTE: can't use not-pure assumeSafeAppend
+        if (buffer.length < length)
+        {
+            buffer.length = length;
+        }
+    }
+
+    static nothrow auto get_cxa_demangle()
+    {
+        import core.sharedlib : SharedLib;
+
+        if (!libCXXABI)
+            return null;
+        static SharedLib* lib;
+        if (!lib)
+            lib = new SharedLib(libCXXABI);
+        if (!lib.handle)
+            return null;
+        return lib.getFun!__cxa_demangle;
+    }
+
+}
+
+enum DemangleCXXStatus
+{
+    invalid,
+    invalidLib,
+    invalidPrefix,
+    memoryFail,
+    invalidMangling,
+    invalidArg,
+    success,
+}
+
+enum prefixCXX = "_Z";
+
+char[] demangleCXX(out DemangleCXXStatus status, const(char)[] buf,
+        DemangleCXX* data = Singleton!DemangleCXX(0)) nothrow pure @trusted
+{
+    status = DemangleCXXStatus.invalid;
+    if (!buf.startsWith(prefixCXX))
+    {
+        status = DemangleCXXStatus.invalidPrefix;
+        return null;
+    }
+
+    if (!data.fun_cxa_demangle)
+    {
+        status = DemangleCXXStatus.invalidLib;
+        return null;
+    }
+
+    // make sure null terminated
+    data.reserve(buf.length + 1);
+    data.buffer[0 .. buf.length] = buf;
+    data.buffer[buf.length] = '\0';
+
+    size_t len = data.buffer2.length;
+    int status2;
+    auto ret = data.fun_cxa_demangle(data.buffer.ptr, data.buffer2.ptr, &len, &status2);
+    data.buffer2 = ret[0 .. len];
+
+    switch (status2)
+    {
+    case 0:
+        {
+            status = DemangleCXXStatus.success;
+            assert(len > 0 && ret[len - 1] == '\0');
+            return ret[0 .. len - 1];
+        }
```

Comment:
You are aware that you don't need the extra scope here?

Reply:
done